### ISSUE TITLE
KeyValue doesn't work on standard entities

### ISSUE THREAD

**Issue by wootguy** [SamVanheer/SC_Angelscript#43]
For example, if you try to set the "map" or "spawnflags" key on a trigger_changelevel, nothing will happen (only tested with plugin script).

[According to the SDK,](https://github.com/ValveSoftware/halflife/blob/5d761709a31ce1e71488f2668321de05f791b405/dlls/triggers.cpp#L1369) it should overwrite the old m_szMapName value.

**Comment by SamVanheer** [comment]
Sven Co-op allows some cvars to override trigger_changelevel, some spawnflags may not be read from after Spawn() has been called. Which spawnflags are you trying to set for which entities?

**Comment by wootguy** [comment]
I was trying to change _anything_ on a trigger_changelevel. The spawnflag I set was 2 (USE only). "targetname" can't be changed either (to test that I tried FireTargets after setting targetname and it didn't trigger). I remember failing with KeyValue a few months back with other entities but can't remember what they were.

If I could read these keyvalues from the entity and make a new one that would be nice too, but it [doesn't look like that's possible](http://forums.svencoop.com/showthread.php/43973-KeyValueBuffer-GetValue-doesn-t-work) (map scripts only maybe?).

**Comment by SamVanheer** [comment]
Ok, looks like you've fallen into another GoldSource trap.

CBaseEntity::KeyValue only sets keyvalues that aren't part of entvars_t. If you want to set those, use CEntityFuncs::DispatchKeyValue, or if you only want to set entvars_t keys, use EntvarsKeyvalue.

DispatchKeyValue will handle any and all keyvalues supported by an entity, including targetname and spawnflags. That's how the engine does it, so you should use it everywhere.

Reading keyvalues isn't supported beyond custom keyvalues. The code for custom keyvalues can be modified to read others, provided the internal GetKeyValue method is overridden to return it. So it's up to the devs to add that. I'd highly recommend overhauling the save/restore code, like i've done with HLEnhanced, so you can use it to get most of them right away.

KeyValueBuffer is intended for player data only. It can read from either the info key buffer, which has client cvars, or the physics key buffer, which is used to share physics settings, like whether the player has the longjump module or not.

**Comment by wootguy** [comment]
Alright then. I've tried DispatchKeyValue and setting the value directly with pEnt.pev.targetname but it still won't respond to FireTargets (or even Use).

I at least know that adding read access to the content directory is on the dev todo list. Worst case scenario I'll read the entity data directly from the bsp.

**Comment by SamVanheer** [comment]
Have you tried setting the name of any other entities and triggering those? trigger_changelevel is a bit of a special case due to how it works. I suspect it checks if the activator is a player before doing anything.

**Comment by wootguy** [comment]
Yeah it does work with other entities. I just tried with env_beam, env_laser, env_spark and func_door but only func_door worked. I'm guessing the other ents didn't toggle because they didn't spawn with that flag enabled.

**Comment by SamVanheer** [comment]
You should ask them to elaborate on how you can successfully trigger a trigger_changelevel. I'm guessing you'll need a player to do it, so if you need to changelevel, try using the "changelevel &lt;mapname&gt;" server command.

**Comment by wootguy** [comment]
In my specific case a changelevel command won't work because I don't know what the next level is (and I can't trust the map cycle). Triggering with a player as the activator and caller didn't work for the changelevel with no targetname. I'll ask about this on the discord channel.
